cpp/src/transform/encode.cpp

```diff
@@ -1,14 +1,56 @@
 #include <cudf/column/column.hpp>
 #include <cudf/column/column_view.hpp>
 #include <cudf/copying.hpp>
+#include <cudf/detail/gather.hpp>
 #include <cudf/detail/search.hpp>
+#include <cudf/detail/sorting.hpp>
 #include <cudf/detail/stream_compaction.hpp>
 #include <cudf/detail/transform.hpp>
 #include <cudf/table/table_view.hpp>
+#include <iostream>
+#include <numeric>
 
 namespace cudf {
 namespace detail {
 
+std::pair<std::unique_ptr<table>, std::unique_ptr<column>> encode(
+  table_view const& input_table, rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  // side effects of this function we are now dependent on:
+  // - resulting table elements are sorted ascending
+
+  std::vector<size_type> columns(input_table.num_columns());
+  std::iota(columns.begin(), columns.end(), 0);
+
+  auto keys_table = cudf::detail::drop_duplicates(
+    input_table, columns, duplicate_keep_option::KEEP_FIRST, null_equality::EQUAL, mr, stream);
+
+  if (cudf::has_nulls(keys_table->view())) {
+    auto sort_order = cudf::detail::sorted_order(
+      keys_table->view(),
+      {},
+      std::vector<null_order>(keys_table->num_columns(), null_order::AFTER),
+      rmm::mr::get_default_resource(),
+      stream);
+    keys_table = cudf::detail::gather(keys_table->view(),
+                                      sort_order->view(),
+                                      cudf::detail::out_of_bounds_policy::FAIL,
+                                      cudf::detail::negative_index_policy::NOT_ALLOWED,
+                                      mr,
+                                      stream);
```

Comment:
Instead of calling `sorted_order()` followed by `gather()` why not just call `sort()`?

Reply:
👍 

After giving this a bit more thought, I'm realizing that maybe the expensive multi-column sort isn't needed. What we really want to do is rearrange the rows according to a "known" gather map. Specifically, the first `n` rows should be pushed to the bottom of the table (in reverse order), where `n` is the number of rows containing null values.